### ISSUE TITLE
[Feature Request] Disable Undo Point Option

### ISSUE THREAD

**Issue by gentlegiantJGC** [Amulet-Team/Amulet-Map-Editor#663]
### The Problem
The undo point operation takes quite a while to run.
Many users have expressed wanting to be able to disable this.

### About
The undo point was added in as a mechanism to allow users to undo changes but also as a mechanism of freeing RAM.

The simple approach to chunk management is store all chunks as they are loaded in a dictionary in the world.
This has the side effect of accumulating a large amount of RAM if a large number of chunks are loaded.
We can unload chunks if they are in the same state as on disk to free a bit of memory but only so much.

The solution was to store the chunk state to disk and reload from that.

After an operation is run we look through all the chunks stored in RAM for that level and check if the chunk has been marked as changed.
If it has changed, we create a backup of it on disk.
Once all changed chunks have been saved, we delete all chunks in RAM and successive calls to get_chunk load from the disk cache.
This has the benefit that if a chunk was changed but not marked as changed those changes will get discarded.

Changing the undo point behaviour is possible but it is currently a rather core part of how Amulet operates so it can't currently just be disabled.

To remove the undo point we would need to disable the disk caching system and all changed chunks would have to exist purely in RAM. We would also need to make sure that the chunks in RAM were not cleared.

If I have more thoughts on this I will modify this.

**Comment by liquidwater2** [comment]
Yes please let us do this!!! I’m trying to move a 47 billion block structure segment by segment and the create undo point takes triple as long as each paste operation. Although I’ve not fully wrapped my head around the ram usage portion of this post I support the concept.

**Comment by Podshot** [comment]
A possible approach we could consider using would be using a local (most likely non-relational) database backend and use that to store the chunk history. The benefits would be that all of the logic of when to off-load old data to disk would be handled by better optimized code for the data we're storing as well as taking advantage that our main threads wouldn't get locked up while the database might be doing it's logic. However, the downsides is that as far as my research has gone into this, it seems most database backends (Redis and MongoDB are the ones I've looked into the most for this) require an separate server process (IE: it can a local server but can't initiated by python (except for running a command-line command) and thus adds something we need to install/bundle that may conflict for a variety of reasons). I'll keep putting some thought into this as well

**Comment by MhikeiMPC** [comment]
but for those of us with lots of ram (i have 32gb) i dont want chunks to prematurely delete themselves from memory, i want to be able to use all the ram i paid for so if this feature were to be considered please also consider those with higher specs.

**Comment by PREMIEREHELL** [comment]
I really think that all that sould be saved is the changes that the user makes, like a ready to apply buffer, that does not apply anything until user saves, upon that user gets asked if he wants an undo or to save a backup of the edited chunks so they can reopen amulet and revert back if it did not go as planned. I think this is a better option because undo does not help after you find a mistake in game. You will not nees to store chunk history only what needs to take place to cause the changes.

**Comment by gentlegiantJGC** [comment]
I think it would make more sense to merge the disk and ram undo options into one and have a configurable amount of chunks/memory until pushing the chunks to deep storage on disk.

**Comment by Buzzb0mber** [comment]
I'm in full support of having the option to disable the undo point entirely. It easily more than doubles the amount of time it takes to commit a change to a map for which I've already made several backup copies, the function has it's uses but for large-scale projects it can be crippling to the point of being an active detriment.

**Comment by PREMIEREHELL** [comment]
I'm also working on plugins at bypass The undo point,  By simply directly editing the world files.  Yet any mistakes are on you.
